### PR TITLE
CPLAT-9205 Codemod for New Boilerplate

### PR DESCRIPTION
**Pull request by sydneyjodon-wk**
## Motivation
New over_react component boilerplate is coming soon, and our consumers deserve an easy way to make use of it!
   
## Changes
- [x] A migrator that acts on simple `@Props()` / `@State()` classes
  (#73)
- [x] A migrator that acts on `@Props()` / `@State()`  classes with advanced inheritance 
  (#76, #79, #81, #82)
- [x] A migrator that acts on standalone `@PropsMixin()` / `@StateMixin()` classes
  (#77)
- [x] A migrator that removes the public _(un-prefixed)_ "companion" classes
  (#75)
- [x] Utilities to determine if a class is public API before performing a migration
  (#74, #80, #83)
- [x] A migrator that removes `@Props`, `@State()`, `@PropsMixin()`, `@StateMixin()`, `@Factory()` and `@Component2()` annotations that are not required by the new boilerplate
  (#78)

#### Release Notes
Add codemod to assist over_react consumers in their migration to the new boilerplate changes.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
      - [ ] Run the boilerplate codemod on web_skin_dart and verify that the codemod updates props and state classes as expected.
        - `pub global run over_react_codemod:boilerplate_upgrade`
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/boilerplate_suggestors/boilerplate_utilities.dart [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+/// Simple means:
+/// - Has no mixins
+/// - Extends from UiProps
+bool isSimplePropsOrStateClass(ClassDeclaration classNode) {
+  // Only validate props or state classes
+  assert(isAPropsOrStateClass(classNode));
+
+  final superClass = classNode.extendsClause.superclass.name.toSource();
+
+  if (superClass != 'UiProps' && superClass != 'UiState') return false;
```
Comment (sydneyjodon-wk) [inline lib/src/boilerplate_suggestors/boilerplate_utilities.dart]: This needs a null check on the `extendsClause`. Without it, there are errors when the codemod is run on `web_skin_dart`.

Maybe change to something like:
```
final superClass = classNode?.extendsClause?.superclass?.name?.toSource?.call();

if (superClass == null || (superClass != 'UiProps' && superClass != 'UiState')) return false;
```

File: lib/src/boilerplate_suggestors/boilerplate_utilities.dart [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  return isAssociatedWithComponent2(node) &&
+      isAPropsOrStateClass(node) &&
+      // Stub while <https://jira.atl.workiva.net/browse/CPLAT-9308> is in progress
+      !_isPublic(node);
+}
+
+/// A simple RegExp against the parent of the class to verify it is `UiProps`
+/// or `UiState`.
+bool extendsFromUiPropsOrUiState(ClassDeclaration classNode) =>
+    classNode.extendsClause.superclass.name
```
Comment (sydneyjodon-wk) [inline lib/src/boilerplate_suggestors/boilerplate_utilities.dart]: This is also causing errors while testing on `web_skin_dart`. I think it just needs a null check.

File: test/boilerplate_suggestors/advanced_props_and_state_class_migrator_test.dart [hunk 3]
```diff
@@ -0,0 +1,2420 @@
...
+            test('with no members', () {
+              const input = r'''
+                  @PropsMixin()
+                  abstract class BreadcrumbPathPropsMixin {}
+                  
+                  @AbstractProps()
+                  class _$AbstractBreadcrumbPathProps extends AbstractGraphFormProps
+                      with
+                          BreadcrumbPathPropsMixin,
+                          // ignore: mixin_of_non_class, undefined_clas
```
Comment (greglittlefield-wf) [inline test/boilerplate_suggestors/advanced_props_and_state_class_migrator_test.dart]: If the ignore comment is at the end of a line, it'll get applied to this line, and not the subsequent line.

So, in this case, `$UnconvertedMixin` would still have an analysis error.

We need this comment to be on its own line:
```suggestion
                with ${publicPropsClassName}Mixin, ConvertedMixin, UnconvertedMixin, 
                // ignore: mixin_of_non_class, undefined_class
                \$UnconvertedMixin;
```
Doesn't matter where the comment goes so long as it has one space of between it and the start of the line; dartfmt will handle it:
```dart
@Props()
class FooProps = UiProps with
 FooPropsMixinn,
 ConvertedMixin,
 UnconvertedMixin,
 // ignore: mixin_of_non_class, undefined_class
 $UnconvertedMixin;
```
Formatted
```dart
@Props()
class FooProps = UiProps
    with
        FooPropsMixinn,
        ConvertedMixin,
        UnconvertedMixin,
        // ignore: mixin_of_non_class, undefined_class
        $UnconvertedMixin;
```

**Comment by joebingham-wk** [comment]
@Workiva/release-management-p

**Pull request merged**